week_3/Binary_Tree/invert_binary_tree/invert_binary_tree.test.js

```diff
@@ -0,0 +1,13 @@
+const invertTree = require("./invert_binary_tree");
+const BinarySearchTree = require("../class");
+
+test("invert_binary_tree works", () => {
+	let bst = new BinarySearchTree();
+	bst.insert(10);
+	bst.insert(15);
+	bst.insert(6);
+	bst.insert(3);
+	bst.insert(8);
+	bst.insert(20);
+	expect(invertTree(bst)).toBe([10, 15, 6, null, 20, 3, 8]);
```

Comment:
is `invertTree` returning an array? It appears to be returning a root node?

Reply:
This is pretty much the crux of my issue. On Leetcode, the "expected" value is given as an array, even though the function is returning a root node.

I see your other comment about how this issue is only designed as a shallow solution and doesn't actually implement a BST class.

My confusion comes from the fact that the Leetcode problem this is taken from (https://leetcode.com/problems/invert-binary-tree/ ), provides the following JSDoc comment: 

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
 
 It also accepts the solution that I wrote, which treats each array element as a "TreeNode" object.

I don't understand why this Object information is provided if a "shallow"/array solution is adequate.

Is it just shorthand for easy readability, while more complex tests are occurring in the background in order to verify the "real" function? If so, should I try to emulate these more complex tests, or rewrite the function as a "shallow"/array solution?

Reply:
Yeah, that's confusing. It appears leetcode's exercise tests, that validate your solution, aren't public. These tests would be a good way of reverse engineering what they're asking for.

In lieu of that, you could construct test data with your BST class and then use it's root node as the input parameter for your test. You could then recreate your test data, but inverted. You'd then implement a function that compares two different BST classes to verify that your expected output matches the output you received.

Reply:
When you say that I'd implement a function that compares two different BST classes, I'm a little confused by what you mean. Do you just mean that I need to check for equivalency between two different BST instances using Jest's "expect" function?

Reply:
I did my best interpretation of this in the test file that I just committed. It's fairly difficult to read and understand, but I'm not sure if there's a better method given the complexity and specificity of the data that I'm working with here. 

The only way that I could formulate the string to "expect" here was to verify elsewhere that the algorithm does in fact work, console.log a stringified JSON object, and then copy that into the "expect" function. This seems like a backwards way to construct a test, although the specific case here might be so particular and impractical as to make the extra effort needed to "truly" test the function overkill. I'm not sure.

Reply:
There's two trains of thought you can apply to generating test data:
1. Creating your own "expect" data can often result in figuring out that your solution was wrong. You'll run the test, see it fail, scratch your head, and dive deep into the implementation; only to eventually realize you were accidentally checking if something was "true" when you should have been checking if that was "false." This happens so frequently that it just happened to me this last Friday 😅.
2. If you use the function itself as the "expect" data, you must take extra time to double and triple check something is working. It's not all that uncommon I'll discover a unit test that is passing when it shouldn't be, for this reason. In very rare cases this is a requirement, but it should avoided at all costs. For data like this, I'd suggest taking the time to construct your test data manually, even if that's a slow process, to ensure everything is correct. If you run your test and see it fail, it might be because of your manually constructed test data being incorrect, but you should never assume that; always pick apart your thought process carefully to ensure you're certain which output is correct.